src/bash/patches/0001-Add-plugin-support-to-bash.patch

```diff
@@ -0,0 +1,793 @@
++#  include <sys/param.h>
++#endif
++
++#if defined (HAVE_UNISTD_H)
++#  include <unistd.h>
++#endif
++
++#include "posixtime.h"
++
++#if defined (HAVE_SYS_RESOURCE_H) && !defined (RLIMTYPE)
++#  include <sys/resource.h>
++#endif
++
++#if defined (HAVE_SYS_TIMES_H) && defined (HAVE_TIMES)
++#  include <sys/times.h>
++#endif
++
++#include <errno.h>
++
++#if !defined (errno)
++extern int errno;
++#endif
++
++#define NEED_FPURGE_DECL
++#define NEED_SH_SETLINEBUF_DECL
++
++#include "bashansi.h"
++#include "bashintl.h"
++
++#include "shell.h"
++#include <y.tab.h>    /* use <...> so we pick it up from the build directory */
++#include "error.h"
++#include "flags.h"
++#include "builtins.h"
++#include "hashlib.h"
++#include "jobs.h"
++#include "execute_cmd.h"
++#include "findcmd.h"
++#include "redir.h"
++#include "trap.h"
++#include "pathexp.h"
++#include "hashcmd.h"
++
++
++#if defined (BASH_SHELL_EXECVE_PLUGIN)
++#include "plugin.h"
++#endif /* BASH_SHELL_EXECVE_PLUGIN */
++
++#if defined (COND_COMMAND)
++#  include "test.h"
++#endif
++
++#include "builtins/common.h"
++
++#include "builtins/getopt.h"
++
++#include <glob/strmatch.h>
++#include <tilde/tilde.h>
++
++#if defined (BUFFERED_INPUT)
++#  include "input.h"
++#endif
++
++#if defined (ALIAS)
++#  include "alias.h"
++#endif
++
++#if defined (HISTORY)
++#  include "bashhist.h"
++#endif
++
++#if defined (HAVE_MBSTR_H) && defined (HAVE_MBSCHR)
++#  include <mbstr.h>        /* mbschr */
++#endif
++
++/* configuration buffer size */
++#define CONFIG_BUFFER_SIZE 256
++
++/* configuration plugin setting key */
++#define CONFIG_PLUGIN_SETTING_KEY "plugin="
++
++/* configuration termination characters */
++#define CONFIG_SETTING_TERMINATIONS " \t\n\r\f"
++
++/* configuration comments start character */
++#define CONFIG_SETTING_COMMENTS_START_CHAR '#'
++
++/* plugin configration file */
++const char *plugin_config_file = "/etc/bash_plugins.conf";
++
++/* plugin on_shell_execve function name */
++static const char *on_shell_execve_function_name = "on_shell_execve";
++
++/* plugin plugin_init function name */
++static const char *plugin_init_function_name = "plugin_init";
++
++/* plugin plugin_uninit function name */
++static const char *plugin_uninit_function_name = "plugin_uninit";
++
++/* plugin handle for test */
++PLUGIN_NODE *global_plugin_list = NULL;
++
++/* Load plugin by plugin path */
++void
++append_plugin(
++        plugin_handle,
++        on_shell_execve,
++        plugin_init,
++        plugin_uninit)
++     void *plugin_handle;
++     on_shell_execve_t *on_shell_execve;
++     plugin_init_t *plugin_init;
++     plugin_uninit_t *plugin_uninit;
++{
++    /* Create and initialize new plugin */
++    PLUGIN_NODE *new_plugin_node = (PLUGIN_NODE*)malloc(sizeof(PLUGIN_NODE));
++    if (new_plugin_node == NULL)
++    {
++        /* When allocate memory failed, stop and return. */
++#ifdef DEBUG
++        itrace("Plugin: failed to allocate memory for plugin node.\n");
++#endif
++        return;
```

Comment:
>return [](http://example.com/codeflow?start=9&length=6)

return nothing, how does caller know malloc failure? how is the caller expected to handle the failure?

Reply:
return error code, and caller output error message when malloc failed.